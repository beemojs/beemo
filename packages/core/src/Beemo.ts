import fs from 'fs-extra';
import path from 'path';
import { Event } from '@boost/event';
import { bool, number, string, shape } from 'optimal';
import CleanupConfigsRoutine from './routines/CleanupConfigsRoutine';
import ResolveConfigsRoutine from './routines/ResolveConfigsRoutine';
import RunDriverRoutine from './routines/RunDriverRoutine';
import Driver from './Driver';
import Context from './contexts/Context';
import DriverContext from './contexts/DriverContext';
import { Argv, Execution, BeemoConfig, BeemoPluginRegistry, DriverOptions } from './types';

export function configBlueprint() {
  return {
    configure: shape({
      cleanup: bool(false),
      parallel: bool(true),
    }),
    execute: shape({
      concurrency: number(),
      graph: bool(true),
    }),
    module: process.env.BEEMO_CONFIG_MODULE
      ? string(process.env.BEEMO_CONFIG_MODULE)
      : string().required(),
  };
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export default class Beemo<T = any> extends Tool<BeemoPluginRegistry, BeemoConfig<T>> {
  onRunDriver = new Event<[DriverContext, Driver<object, DriverOptions>]>('run-driver');

  constructor(argv: Argv, binName?: string, testingOnly: boolean = false) {
    super(
      {
        appName: 'beemo',
        appPath: path.join(__dirname, '..'),
        configBlueprint: configBlueprint(),
        configName: binName || 'beemo',
        scoped: true,
      },
      argv,
    );

    // Abort early for testing purposes
    if (testingOnly) {
      return;
    }

    this.initialize();

    // Set footer after messages have been loaded
    const footer = this.msg('app:poweredBy', { version });

    this.options.footer = `\n${this.isCI() ? '' : 'ðŸ¤–  '}${footer}`;
  }

  /**
   * Execute all routines for the chosen driver.
   */
  async runDriver(
    args: DriverContext['args'],
    driverName: string,
    parallelArgv: Argv[] = [],
  ): Promise<Execution[]> {
    const driver = this.getPlugin('driver', driverName);
    const context = this.prepareContext(new DriverContext(args, driver, parallelArgv));
    const version = driver.getVersion();

    this.onRunDriver.emit([context, driver], driverName);

    this.debug('Running with %s v%s driver', driverName, version);

    const pipeline = this.startPipeline(context)
      .pipe(new ResolveConfigsRoutine('config', this.msg('app:configGenerate')))
      .pipe(
        new RunDriverRoutine(
          'driver',
          this.msg('app:driverRun', {
            name: driver.metadata.title,
            version,
          }),
        ),
      );

    // Only add cleanup routine if we need it
    if (this.config.configure.cleanup) {
      pipeline.pipe(new CleanupConfigsRoutine('cleanup', this.msg('app:cleanup')));
    }

    return pipeline.run(driverName);
  }

  /**
   * Delete config files if a process fails.
   */
  private handleCleanupOnFailure(code: number, context: Context) {
    if (code === 0) {
      return;
    }

    // Must not be async!
    if (Array.isArray(context.configPaths)) {
      context.configPaths.forEach((config) => {
        fs.removeSync(config.path.path());
      });
    }
  }
}
