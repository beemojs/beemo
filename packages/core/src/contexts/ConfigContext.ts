import { Driver } from '../Driver';
import { Context } from './Context';

export class ConfigContext<O extends object = {}> extends Context<O, string[]> {
	// List of drivers involved in the current pipeline
	drivers: Set<Driver> = new Set();

	/**
	 * Add a driver as a dependency.
	 */
	addDriverDependency(driver: Driver): this {
		if (driver instanceof Driver) {
			this.drivers.add(driver);
		} else {
			throw new TypeError('Invalid driver. Must be an instance of `Driver`.');
		}

		return this;
	}
}
