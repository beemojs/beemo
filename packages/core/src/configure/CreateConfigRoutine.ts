import chalk from 'chalk';
import fs from 'fs-extra';
import path from 'path';
import camelCase from 'lodash/camelCase';
import { ConfigLoader, Routine, Predicates } from '@boost/core';
import Driver from '../Driver';
import ConfigContext from '../contexts/ConfigContext';
import { STRATEGY_COPY, STRATEGY_REFERENCE, STRATEGY_CREATE, STRATEGY_NATIVE } from '../constants';
import { BeemoTool } from '../types';

export interface ConfigObject {
  [key: string]: any;
}

export interface CreateConfigOptions {
  driver: Driver;
}

export default class CreateConfigRoutine<T extends ConfigContext> extends Routine<
  T,
  BeemoTool,
  CreateConfigOptions
> {
  blueprint({ instance }: Predicates) /* infer */ {
    return {
      driver: instance(Driver, true)
        .required()
        .notNullable(),
    };
  }

  bootstrap() {
    const { tool } = this;
    const { metadata, options } = this.options.driver;
    const name = metadata.title;
    const strategy =
      options.strategy === STRATEGY_NATIVE ? metadata.configStrategy : options.strategy;

    this.task(tool.msg('app:configSetEnvVars', { name }), this.setEnvVars);

    switch (strategy) {
      case STRATEGY_REFERENCE:
        this.task(tool.msg('app:configReference', { name }), this.referenceConfigFile);
        break;

      case STRATEGY_COPY:
        this.task(tool.msg('app:configCopy', { name }), this.copyConfigFile);
        break;

      case STRATEGY_CREATE:
        this.task(tool.msg('app:configCreateLoadFile', { name }), this.loadConfigFromSources);
        this.task(tool.msg('app:configCreateLoadPackage', { name }), this.extractConfigFromPackage);
        this.task(tool.msg('app:configCreateMerge', { name }), this.mergeConfigs);
        this.task(tool.msg('app:configCreate', { name }), this.createConfigFile);
        break;

      default:
        this.skip(true);
        break;
    }
  }

  execute(): Promise<string> {
    return this.serializeTasks([]);
  }

  /**
   * Copy configuration file from module.
   */
  async copyConfigFile(context: ConfigContext): Promise<string> {
    const { metadata, name } = this.options.driver;
    const configLoader = new ConfigLoader(this.tool);
    const sourcePath = this.getConfigPath(configLoader);
    const configPath = path.join(context.cwd, metadata.configName);

    if (!sourcePath) {
      throw new Error(this.tool.msg('errors:configCopySourceMissing'));
    }

    const config = this.loadConfig(configLoader, sourcePath);

    this.debug('Copying config file to %s', chalk.cyan(configPath));

    this.options.driver.config = config;

    context.addConfigPath(name, configPath);

    this.tool.emit(`${name}.copy-config-file`, [context, configPath, config]);

    return fs
      .copy(sourcePath, configPath, {
        overwrite: true,
      })
      .then(() => configPath);
  }

  /**
   * Create a temporary configuration file or pass as an option.
   */
  async createConfigFile(context: ConfigContext, config: object): Promise<string> {
    const { metadata, name } = this.options.driver;
    const configPath = path.join(context.cwd, metadata.configName);

    this.debug('Creating config file %s', chalk.cyan(configPath));

    this.options.driver.config = config as any;

    context.addConfigPath(name, configPath);

    this.tool.emit(`${name}.create-config-file`, [context, configPath, config]);

    return fs
      .writeFile(configPath, this.options.driver.formatConfig(config))
      .then(() => configPath);
  }

  /**
   * Extract configuration from "beemo.<driver>" within the local project's package.json.
   */
  extractConfigFromPackage(
    context: ConfigContext,
    prevConfigs: ConfigObject[],
  ): Promise<ConfigObject[]> {
    const { name } = this.options.driver;
    const { config } = this.tool;
    const configs = [...prevConfigs];
    const configName = this.getConfigName(name);

    this.debug.invariant(
      !!config[configName],
      `Extracting ${chalk.green(name)} config from package.json`,
      'Exists, extracting',
      'Does not exist, skipping',
    );

    if (config[configName]) {
      const pkgConfig = config[configName];

      configs.push(pkgConfig);

      this.tool.emit(`${name}.load-package-config`, [context, pkgConfig]);
    }

    return Promise.resolve(configs);
  }

  /**
   * Return file name camel cased.
   */
  getConfigName(name: string): string {
    return camelCase(name);
  }

  /**
   * Return absolute file path for config file within configuration module,
   * or an empty string if it does not exist.
   */
  getConfigPath(configLoader: ConfigLoader, forceLocal: boolean = false): string {
    const { cwd, workspaceRoot } = this.context;
    const moduleName = this.tool.config.module;
    const { name } = this.options.driver;
    const configName = this.getConfigName(name);
    const isLocal = moduleName === '@local' || forceLocal;

    // Allow for local development
    const filePath = isLocal
      ? path.join(workspaceRoot || cwd, `configs/${configName}.js`)
      : configLoader.resolveModuleConfigPath(configName, moduleName);
    const fileExists = fs.existsSync(filePath);

    this.debug.invariant(
      fileExists,
      isLocal
        ? `Loading ${chalk.green(name)} config from local consumer`
        : `Loading ${chalk.green(name)} config from configuration module ${chalk.yellow(
            moduleName,
          )}`,
      'Exists, loading',
      'Does not exist, skipping',
    );

    return fileExists ? filePath : '';
  }

  /**
   * Merge multiple configuration sources using the current driver.
   */
  mergeConfigs(context: ConfigContext, configs: ConfigObject[]): Promise<ConfigObject> {
    const { name } = this.options.driver;

    this.debug('Merging %s config from %d sources', chalk.green(name), configs.length);

    const config = configs.reduce(
      (masterConfig, cfg) => this.options.driver.mergeConfig(masterConfig, cfg),
      {},
    );

    this.tool.emit(`${name}.merge-config`, [context, config]);

    return Promise.resolve(config);
  }

  /**
   * Load a config file with passing the args and tool to the file.
   */
  loadConfig(configLoader: ConfigLoader, filePath: string): ConfigObject {
    const config = configLoader.parseFile(filePath, [], { errorOnFunction: true });

    this.tool.emit(`${this.options.driver.name}.load-module-config`, [
      this.context,
      filePath,
      config,
    ]);

    return config;
  }

  /**
   * Load config from the provider configuration module
   * and from the local configs/ folder in the consumer.
   */
  loadConfigFromSources(
    context: ConfigContext,
    prevConfigs: ConfigObject[],
  ): Promise<ConfigObject[]> {
    const configLoader = new ConfigLoader(this.tool);
    const modulePath = this.getConfigPath(configLoader);
    const localPath = this.getConfigPath(configLoader, true);
    const configs = [...prevConfigs];

    if (modulePath) {
      configs.push(this.loadConfig(configLoader, modulePath));
    }

    // Local files should override anything defined in the configuration module above
    // Also don't double load files, so check against @local to avoid
    if (localPath && localPath !== modulePath) {
      configs.push(this.loadConfig(configLoader, localPath));
    }

    return Promise.resolve(configs);
  }

  /**
   * Reference configuration file from module using a require statement.
   */
  referenceConfigFile(context: ConfigContext): Promise<string> {
    const { metadata, name } = this.options.driver;
    const configLoader = new ConfigLoader(this.tool);
    const sourcePath = this.getConfigPath(configLoader);
    const configPath = path.join(context.cwd, metadata.configName);

    if (!sourcePath) {
      throw new Error(this.tool.msg('errors:configReferenceSourceMissing'));
    }

    const config = this.loadConfig(configLoader, sourcePath);

    this.debug('Referencing config file to %s', chalk.cyan(configPath));

    this.options.driver.config = config;

    context.addConfigPath(name, configPath);

    this.tool.emit(`${name}.reference-config-file`, [context, configPath, config]);

    return fs
      .writeFile(
        configPath,
        `module.exports = require('./${path.relative(context.cwd, sourcePath)}');`,
      )
      .then(() => configPath);
  }

  /**
   * Set environment variables defined by the driver.
   */
  setEnvVars(context: ConfigContext, configs: ConfigObject[]): Promise<any> {
    const { env } = this.options.driver.options;

    // TODO: This may cause collisions, isolate in a child process?
    Object.keys(env).forEach(key => {
      process.env[key] = env[key];
    });

    return Promise.resolve(configs);
  }
}
