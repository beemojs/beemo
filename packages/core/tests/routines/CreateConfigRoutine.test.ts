import fs from 'fs-extra';
import ConfigLoader from '@boost/core/lib/ConfigLoader';
import Beemo from '../../src/Beemo';
import CreateConfigRoutine from '../../src/routines/CreateConfigRoutine';
import Driver from '../../src/Driver';
import {
  STRATEGY_COPY,
  STRATEGY_REFERENCE,
  STRATEGY_CREATE,
  STRATEGY_NONE,
  STRATEGY_NATIVE,
} from '../../src/constants';
import {
  stubConfigContext,
  mockDebugger,
  mockTool,
  prependRoot,
  mockDriver,
} from '../../src/testUtils';
import ConfigContext from '../../src/contexts/ConfigContext';

jest.mock('@boost/core/lib/ConfigLoader');

describe('CreateConfigRoutine', () => {
  const oldExistsSync = fs.existsSync;
  const oldWriteFile = fs.writeFile;
  const oldCopy = fs.copy;
  let routine: CreateConfigRoutine<ConfigContext>;
  let driver: Driver;
  let tool: Beemo;

  beforeEach(() => {
    tool = mockTool();

    driver = mockDriver(
      'babel',
      tool,
      {
        configName: 'babel.config.js',
        configOption: '--config-file',
        title: 'Babel',
      },
      false,
    );
    driver.configure({ args: ['--qux'] });
    driver.bootstrap();

    routine = new CreateConfigRoutine('babel', 'Configure Babel', { driver });
    routine.context = stubConfigContext();
    routine.tool = tool;
    routine.tool.config.module = '@local';
    routine.debug = mockDebugger();

    jest.spyOn(fs, 'existsSync').mockImplementation(() => true);
    // eslint-disable-next-line @typescript-eslint/no-misused-promises
    jest.spyOn(fs, 'writeFile').mockImplementation(() => Promise.resolve());
    // eslint-disable-next-line @typescript-eslint/no-misused-promises
    jest.spyOn(fs, 'copy').mockImplementation(() => Promise.resolve());

    // @ts-ignore
    ConfigLoader.mockImplementation(() => ({
      parseFile: () => ({ foo: 123 }),
    }));

    process.beemo = {
      context: routine.context,
      tool,
    };
  });

  afterEach(() => {
    // @ts-ignore
    ConfigLoader.mockReset();

    delete process.beemo;

    fs.existsSync = oldExistsSync;
    fs.writeFile = oldWriteFile;
    fs.copy = oldCopy;
  });

  describe('constructor()', () => {
    it('errors if no driver defined', () => {
      expect(() => {
        routine = new CreateConfigRoutine('title', 'title');
      }).toThrowErrorMatchingSnapshot();
    });
  });

  describe('bootstrap()', () => {
    describe('create strategy', () => {
      beforeEach(() => {
        driver.configure({ strategy: STRATEGY_CREATE });
      });

      it('sets the correct tasks', () => {
        routine.bootstrap();

        expect(routine.tasks).toHaveLength(5);
        expect(routine.tasks[4].title).toBe('Creating Babel config file');
      });
    });

    describe('copy strategy', () => {
      beforeEach(() => {
        driver.configure({ strategy: STRATEGY_COPY });
      });

      it('sets the correct tasks', () => {
        routine.bootstrap();

        expect(routine.tasks).toHaveLength(2);
        expect(routine.tasks[1].title).toBe('Copying Babel config file');
      });
    });

    describe('reference strategy', () => {
      beforeEach(() => {
        driver.configure({ strategy: STRATEGY_REFERENCE });
      });

      it('sets the correct tasks', () => {
        routine.bootstrap();

        expect(routine.tasks).toHaveLength(2);
        expect(routine.tasks[1].title).toBe('Referencing Babel config file');
      });
    });

    describe('native strategy', () => {
      beforeEach(() => {
        driver.configure({ strategy: STRATEGY_NATIVE });
        driver.metadata.configStrategy = STRATEGY_COPY;
      });

      it('uses strategy defined in driver metadata', () => {
        routine.bootstrap();

        expect(routine.tasks).toHaveLength(2);
        expect(routine.tasks[1].title).toBe('Copying Babel config file');
      });
    });

    describe('none strategy', () => {
      beforeEach(() => {
        driver.configure({ strategy: STRATEGY_NONE });
      });

      it('sets no tasks and skips routine', () => {
        routine.bootstrap();

        expect(routine.tasks).toHaveLength(1); // Env vars
        expect(routine.isSkipped()).toBe(true);
      });
    });
  });

  describe('run()', () => {
    beforeEach(() => {
      driver.configure({ strategy: STRATEGY_NATIVE });
    });

    describe('create strategy', () => {
      beforeEach(() => {
        routine.tool.config.babel = { bar: 'abc' };
      });

      it('creates config file if `configStrategy` metadata is create', async () => {
        driver.metadata.configStrategy = STRATEGY_CREATE;

        const envSpy = jest.spyOn(routine, 'setEnvVars');
        const loadSpy = jest.spyOn(routine, 'loadConfigFromSources');
        const extractSpy = jest.spyOn(routine, 'extractConfigFromPackage');
        const mergeSpy = jest.spyOn(routine, 'mergeConfigs');
        const createSpy = jest.spyOn(routine, 'createConfigFile');

        routine.bootstrap();

        const path = await routine.run(routine.context, []);

        expect(envSpy).toHaveBeenCalledWith(routine.context, [], expect.anything());
        expect(loadSpy).toHaveBeenCalledWith(routine.context, [], expect.anything());
        expect(extractSpy).toHaveBeenCalledWith(routine.context, [{ foo: 123 }], expect.anything());
        expect(mergeSpy).toHaveBeenCalledWith(
          routine.context,
          [{ foo: 123 }, { bar: 'abc' }],
          expect.anything(),
        );
        expect(createSpy).toHaveBeenCalledWith(
          routine.context,
          { foo: 123, bar: 'abc' },
          expect.anything(),
        );
        expect(path).toBe(prependRoot('/babel.config.js').path());
        expect(fs.writeFile).toHaveBeenCalledWith(
          prependRoot('/babel.config.js').path(),
          `module.exports = ${JSON.stringify({ foo: 123, bar: 'abc' }, null, 2)};`,
        );
      });

      it('creates config file if `strategy` option is create', async () => {
        driver.configure({ strategy: STRATEGY_CREATE });

        const envSpy = jest.spyOn(routine, 'setEnvVars');
        const loadSpy = jest.spyOn(routine, 'loadConfigFromSources');
        const extractSpy = jest.spyOn(routine, 'extractConfigFromPackage');
        const mergeSpy = jest.spyOn(routine, 'mergeConfigs');
        const createSpy = jest.spyOn(routine, 'createConfigFile');

        routine.bootstrap();

        const path = await routine.run(routine.context, []);

        expect(envSpy).toHaveBeenCalledWith(routine.context, [], expect.anything());
        expect(loadSpy).toHaveBeenCalledWith(routine.context, [], expect.anything());
        expect(extractSpy).toHaveBeenCalledWith(routine.context, [{ foo: 123 }], expect.anything());
        expect(mergeSpy).toHaveBeenCalledWith(
          routine.context,
          [{ foo: 123 }, { bar: 'abc' }],
          expect.anything(),
        );
        expect(createSpy).toHaveBeenCalledWith(
          routine.context,
          { foo: 123, bar: 'abc' },
          expect.anything(),
        );
        expect(path).toBe(prependRoot('/babel.config.js').path());
        expect(fs.writeFile).toHaveBeenCalledWith(
          prependRoot('/babel.config.js').path(),
          `module.exports = ${JSON.stringify({ foo: 123, bar: 'abc' }, null, 2)};`,
        );
      });
    });

    describe('copy strategy', () => {
      it('copies config file if `configStrategy` metadata is copy', async () => {
        driver.metadata.configStrategy = STRATEGY_COPY;

        const envSpy = jest.spyOn(routine, 'setEnvVars');
        const createSpy = jest.spyOn(routine, 'createConfigFile');
        const copySpy = jest.spyOn(routine, 'copyConfigFile');

        routine.bootstrap();

        const path = await routine.run(routine.context, []);

        expect(envSpy).toHaveBeenCalledWith(routine.context, [], expect.anything());
        expect(copySpy).toHaveBeenCalledWith(routine.context, [], expect.anything());
        expect(createSpy).not.toHaveBeenCalled();
        expect(path).toBe(prependRoot('/babel.config.js').path());
      });

      it('copies config file if `strategy` option is copy', async () => {
        driver.configure({ strategy: STRATEGY_COPY });

        const envSpy = jest.spyOn(routine, 'setEnvVars');
        const createSpy = jest.spyOn(routine, 'createConfigFile');
        const copySpy = jest.spyOn(routine, 'copyConfigFile');

        routine.bootstrap();

        const path = await routine.run(routine.context, []);

        expect(envSpy).toHaveBeenCalledWith(routine.context, [], expect.anything());
        expect(copySpy).toHaveBeenCalledWith(routine.context, [], expect.anything());
        expect(createSpy).not.toHaveBeenCalled();
        expect(path).toBe(prependRoot('/babel.config.js').path());
      });
    });

    describe('reference strategy', () => {
      it('references config file if `configStrategy` metadata is reference', async () => {
        driver.metadata.configStrategy = STRATEGY_REFERENCE;

        const envSpy = jest.spyOn(routine, 'setEnvVars');
        const createSpy = jest.spyOn(routine, 'createConfigFile');
        const refSpy = jest.spyOn(routine, 'referenceConfigFile');

        routine.bootstrap();

        const path = await routine.run(routine.context, []);

        expect(envSpy).toHaveBeenCalledWith(routine.context, [], expect.anything());
        expect(refSpy).toHaveBeenCalledWith(routine.context, [], expect.anything());
        expect(createSpy).not.toHaveBeenCalled();
        expect(path).toBe(prependRoot('/babel.config.js').path());
      });

      it('references config file if `strategy` option is reference', async () => {
        driver.configure({ strategy: STRATEGY_REFERENCE });

        const envSpy = jest.spyOn(routine, 'setEnvVars');
        const createSpy = jest.spyOn(routine, 'createConfigFile');
        const refSpy = jest.spyOn(routine, 'referenceConfigFile');

        routine.bootstrap();

        const path = await routine.run(routine.context, []);

        expect(envSpy).toHaveBeenCalledWith(routine.context, [], expect.anything());
        expect(refSpy).toHaveBeenCalledWith(routine.context, [], expect.anything());
        expect(createSpy).not.toHaveBeenCalled();
        expect(path).toBe(prependRoot('/babel.config.js').path());
      });
    });

    describe('native strategy', () => {
      it('falls back to `configStrategy` metadata when `strategy` option is native', async () => {
        driver.configure({ strategy: STRATEGY_NATIVE });
        driver.metadata.configStrategy = STRATEGY_COPY;

        const envSpy = jest.spyOn(routine, 'setEnvVars');
        const createSpy = jest.spyOn(routine, 'createConfigFile');
        const copySpy = jest.spyOn(routine, 'copyConfigFile');

        routine.bootstrap();

        const path = await routine.run(routine.context, []);

        expect(envSpy).toHaveBeenCalledWith(routine.context, [], expect.anything());
        expect(copySpy).toHaveBeenCalledWith(routine.context, [], expect.anything());
        expect(createSpy).not.toHaveBeenCalled();
        expect(path).toBe(prependRoot('/babel.config.js').path());
      });
    });

    describe('none strategy', () => {
      it('does nothing since routine is skipped', async () => {
        driver.configure({ strategy: STRATEGY_NONE });

        const envSpy = jest.spyOn(routine, 'setEnvVars');

        routine.bootstrap();

        await routine.run(routine.context, []);

        expect(envSpy).not.toHaveBeenCalled();
      });
    });
  });

  describe('copyConfigFile()', () => {
    it('adds path to context', async () => {
      await routine.copyConfigFile(routine.context);

      expect(routine.context.configPaths).toEqual([
        { driver: 'babel', path: prependRoot('/babel.config.js') },
      ]);
    });

    it('copies file', async () => {
      const path = await routine.copyConfigFile(routine.context);

      expect(fs.copy).toHaveBeenCalledWith(
        prependRoot('/configs/babel.js').path(),
        prependRoot('/babel.config.js').path(),
        { overwrite: true },
      );
      expect(path).toBe(prependRoot('/babel.config.js').path());
    });

    it('sets config on driver', async () => {
      await routine.copyConfigFile(routine.context);

      expect(driver.config).toEqual({ foo: 123 });
    });

    it('emits `onCopyConfigFile` event', async () => {
      const spy = jest.fn();

      driver.onCopyConfigFile.listen(spy);

      await routine.copyConfigFile(routine.context);

      expect(spy).toHaveBeenCalledWith(routine.context, prependRoot('/babel.config.js').path(), {
        foo: 123,
      });
    });

    it('errors if no source file', async () => {
      routine.getConfigPath = () => null;

      await expect(routine.copyConfigFile(routine.context)).rejects.toThrowErrorMatchingSnapshot();
    });
  });

  describe('createConfigFile()', () => {
    it('adds path to context', async () => {
      await routine.createConfigFile(routine.context, { foo: 'bar' });

      expect(routine.context.configPaths).toEqual([
        { driver: 'babel', path: prependRoot('/babel.config.js') },
      ]);
    });

    it('writes and formats file', async () => {
      const path = await routine.createConfigFile(routine.context, { foo: 'bar' });

      expect(fs.writeFile).toHaveBeenCalledWith(
        prependRoot('/babel.config.js').path(),
        'module.exports = {\n  "foo": "bar"\n};',
      );
      expect(path).toBe(prependRoot('/babel.config.js').path());
    });

    it('sets config on driver', async () => {
      await routine.createConfigFile(routine.context, { foo: 'bar' });

      expect(driver.config).toEqual({ foo: 'bar' });
    });

    it('emits `onCreateConfigFile` event', async () => {
      const spy = jest.fn();

      driver.onCreateConfigFile.listen(spy);

      await routine.createConfigFile(routine.context, { foo: 'bar' });

      expect(spy).toHaveBeenCalledWith(routine.context, prependRoot('/babel.config.js').path(), {
        foo: 'bar',
      });
    });
  });

  describe('extractConfigFromPackage()', () => {
    it('does nothing if config does not exist', async () => {
      const configs = await routine.extractConfigFromPackage(routine.context, []);

      expect(configs).toEqual([]);
    });

    it('extracts config by name', async () => {
      routine.tool.config.babel = { foo: 'bar' };

      const configs = await routine.extractConfigFromPackage(routine.context, []);

      expect(configs).toEqual([{ foo: 'bar' }]);
    });

    it('emits `onLoadPackageConfig` event', async () => {
      const spy = jest.fn();

      driver.onLoadPackageConfig.listen(spy);

      routine.tool.config.babel = { foo: 'bar' };

      await routine.extractConfigFromPackage(routine.context, []);

      expect(spy).toHaveBeenCalledWith(routine.context, { foo: 'bar' });
    });

    it('doesnt trigger `onLoadPackageConfig` if no config', async () => {
      const spy = jest.fn();

      driver.onLoadPackageConfig.listen(spy);

      await routine.extractConfigFromPackage(routine.context, []);

      expect(spy).not.toHaveBeenCalled();
    });
  });

  describe('mergeConfigs()', () => {
    it('merges multiple sources', async () => {
      const config = await routine.mergeConfigs(routine.context, [
        { foo: 123, qux: true },
        { bar: 'abc' },
        { foo: 456 },
      ]);

      expect(config).toEqual({
        foo: 456,
        bar: 'abc',
        qux: true,
      });
    });

    it('calls `mergeConfig` on driver', async () => {
      const spy = jest.spyOn(driver, 'mergeConfig');

      await routine.mergeConfigs(routine.context, [
        { foo: 123, qux: true },
        { bar: 'abc' },
        { foo: 456 },
      ]);

      expect(spy).toHaveBeenCalledTimes(3);
    });

    it('emits `onMergeConfig` event with final config object', async () => {
      const spy = jest.fn();

      driver.onMergeConfig.listen(spy);

      const config = await routine.mergeConfigs(routine.context, [
        { foo: 123, qux: true },
        { bar: 'abc' },
        { foo: 456 },
      ]);

      expect(spy).toHaveBeenCalledWith(routine.context, config);
    });
  });

  describe('loadConfigFromSources()', () => {
    let parseSpy: jest.Mock;

    beforeEach(() => {
      parseSpy = jest.fn();

      // @ts-ignore
      ConfigLoader.mockImplementation(() => ({
        resolveModuleConfigPath: (name: string, moduleName: string) =>
          `/node_modules/${moduleName}/configs/${name}.js`,
        parseFile: parseSpy.mockImplementation(filePath => ({ filePath })),
      }));
    });

    afterEach(() => {
      // @ts-ignore
      ConfigLoader.mockReset();
    });

    it('loads config if it exists', async () => {
      const configs = await routine.loadConfigFromSources(routine.context, []);

      expect(configs).toEqual([{ filePath: prependRoot('/configs/babel.js').path() }]);
    });

    it('does nothing if config does not exist', async () => {
      (fs.existsSync as jest.Mock).mockImplementation(() => false);

      const configs = await routine.loadConfigFromSources(routine.context, []);

      expect(configs).toEqual([]);
    });

    it('uses local path when using @local config', async () => {
      const configs = await routine.loadConfigFromSources(routine.context, []);

      expect(configs).toEqual([{ filePath: prependRoot('/configs/babel.js').path() }]);
    });

    it('uses module path when using custom config', async () => {
      routine.tool.config.module = 'foo-bar';

      const configs = await routine.loadConfigFromSources(routine.context, []);

      expect(configs).toEqual([
        { filePath: '/node_modules/foo-bar/configs/babel.js' },
        { filePath: prependRoot('/configs/babel.js').path() },
      ]);
    });

    it('emits `onLoadModuleConfig` event', async () => {
      const spy = jest.fn();

      driver.onLoadModuleConfig.listen(spy);

      await routine.loadConfigFromSources(routine.context, []);

      expect(spy).toHaveBeenCalledWith(routine.context, prependRoot('/configs/babel.js').path(), {
        filePath: prependRoot('/configs/babel.js').path(),
      });
    });

    it('doesnt trigger `onLoadModuleConfig` event if files does not exist', async () => {
      const spy = jest.fn();

      driver.onLoadModuleConfig.listen(spy);

      (fs.existsSync as jest.Mock).mockImplementation(() => false);

      await routine.loadConfigFromSources(routine.context, []);

      expect(spy).not.toHaveBeenCalled();
    });

    it('parses file', async () => {
      await routine.loadConfigFromSources(routine.context, []);

      expect(parseSpy).toHaveBeenCalledWith(prependRoot('/configs/babel.js').path(), [], {
        errorOnFunction: true,
      });
    });
  });

  describe('referenceConfigFile()', () => {
    it('adds path to context', async () => {
      await routine.referenceConfigFile(routine.context);

      expect(routine.context.configPaths).toEqual([
        { driver: 'babel', path: prependRoot('/babel.config.js') },
      ]);
    });

    it('references file', async () => {
      const path = await routine.referenceConfigFile(routine.context);

      expect(fs.writeFile).toHaveBeenCalledWith(
        prependRoot('/babel.config.js').path(),
        "module.exports = require('./configs/babel.js');",
      );
      expect(path).toBe(prependRoot('/babel.config.js').path());
    });

    it('sets config on driver', async () => {
      await routine.referenceConfigFile(routine.context);

      expect(driver.config).toEqual({ foo: 123 });
    });

    it('emits `onReferenceConfigFile` event', async () => {
      const spy = jest.fn();

      driver.onReferenceConfigFile.listen(spy);

      await routine.referenceConfigFile(routine.context);

      expect(spy).toHaveBeenCalledWith(routine.context, prependRoot('/babel.config.js').path(), {
        foo: 123,
      });
    });

    it('errors if no source file', () => {
      routine.getConfigPath = () => null;

      expect(() => {
        routine.referenceConfigFile(routine.context);
      }).toThrowErrorMatchingSnapshot();
    });
  });

  describe('setEnvVars()', () => {
    it('sets env vars', () => {
      expect(process.env.BEEMO_TEST_VAR).toBeUndefined();

      driver.options.env.BEEMO_TEST_VAR = 'true';

      routine.setEnvVars(routine.context, []);

      expect(process.env.BEEMO_TEST_VAR).toBe('true');
    });
  });
});
