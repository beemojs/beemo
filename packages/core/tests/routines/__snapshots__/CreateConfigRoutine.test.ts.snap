// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateConfigRoutine constructor() errors if no driver defined 1`] = `
"The following validations have failed for \`CreateConfigRoutine\`:
  - Invalid field \\"driver\\". Field is required and must be defined.
  - Invalid field \\"tool\\". Field is required and must be defined."
`;

exports[`CreateConfigRoutine copyConfigFile() errors if no source file 1`] = `"Cannot copy configuration file. Source file does not exist in configuration module."`;

exports[`CreateConfigRoutine referenceConfigFile() errors if no source file 1`] = `"Cannot reference configuration file. Source file does not exist in configuration module."`;
